### ISSUE TITLE
BUG: stats.nbinom fails if n=0 

### ISSUE THREAD

**Issue by pbrod** [scipy/scipy#18974]
### Describe your issue.

The negative binomial distribution with parameters r and p fails to calculate the pmf or cdf for the corner case when n=0.
The pmf is defined recursively as 
 
  $P(k+1) = \frac{(1-p) P(k) (n+k)}{k+1}$ where  $P(0) = p^n$

Thus if n=0 the pmf should give $P(0)=1$ and $P(k)=0$ for $k>0$ whereas 
the scipy.stats implementation only returns NaNs.

### Reproducing Code Example

```python
import scipy.stats as ss
import numpy as np

ss.nbinom(n=0, p=1).pmf(np.arange(3))
ss.nbinom(n=0, p=0.5).pmf(np.arange(3))
ss.nbinom(n=0, p=0.1).cdf(np.arange(3))
```

### Error message

```shell
array([nan, nan, nan])
```

### SciPy/NumPy/Python version and system information

```shell
1.10.0 1.23.5 sys.version_info(major=3, minor=10, micro=9, releaselevel='final', serial=0)
lapack_mkl_info:
    libraries = ['mkl_rt']
    library_dirs = ['C:/Users/pab/AppData/Local/anaconda3_2023\\Library\\lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['C:/Users/pab/AppData/Local/anaconda3_2023\\Library\\include']
lapack_opt_info:
    libraries = ['mkl_rt']
    library_dirs = ['C:/Users/pab/AppData/Local/anaconda3_2023\\Library\\lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['C:/Users/pab/AppData/Local/anaconda3_2023\\Library\\include']
blas_mkl_info:
    libraries = ['mkl_rt']
    library_dirs = ['C:/Users/pab/AppData/Local/anaconda3_2023\\Library\\lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['C:/Users/pab/AppData/Local/anaconda3_2023\\Library\\include']
blas_opt_info:
    libraries = ['mkl_rt']
    library_dirs = ['C:/Users/pab/AppData/Local/anaconda3_2023\\Library\\lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['C:/Users/pab/AppData/Local/anaconda3_2023\\Library\\include']
Supported SIMD extensions in this NumPy install:
    baseline = SSE,SSE2,SSE3
    found = SSSE3,SSE41,POPCNT,SSE42,AVX,F16C,FMA3,AVX2
    not found = AVX512F,AVX512CD,AVX512_SKX,AVX512_CLX,AVX512_CNL
```

**Comment by mdhaber** [comment]
I looked at a few references. Do you have one in mind? Since this is an edge case and others have thought carefully about it, I think we should follow the recommendations of those who have thought more carefully about it.

Some define the distribution only for $n > 0$. SciPy's `nbinom` relies on Boost, which extends the definition to strictly positive $n$. This generalization is referred to as the Polya distribution. It looks like it can be extended to $n = 0$, and Mathematica does, although it is not totally consistent about it.

<details>
You can see that the PDF defined below has $P(X = x) = 0$ for $x \geq 0$, yet the plot shows that the value is $1$ at $x=0$.
![image](https://github.com/scipy/scipy/assets/6570539/ff05ed2f-cba5-4f49-9734-71920bc5fc81)

https://www.wolframalpha.com/input?i=NegativeBinomialDistribution%5B0%2C+0.5%5D

Also from their documentation of `NegativeBinomialDistribution`, they exclude the possibility of $n=0$:
![image](https://github.com/scipy/scipy/assets/6570539/aea7e551-d180-4180-8534-ae36532ad62f)

</details>

My guess is that Boost *chose* not to provide a definition for $n = 0$. We can ask them to reconsider it, but it's not immediately obvious that it is *incorrect* not to define this distribution for $n=0$, so I'd consider this an enhancement (and maybe documentation) issue.

**Comment by pbrod** [comment]
Most references seems to define the negative binomial distribution for $n>0$.
However, the [moment generating function ](https://en.wikipedia.org/wiki/Moment-generating_function)(MGF) for the negative binomial distribution is 
$M_X(t) = \left(\frac{p}{1-(1-p) e^t}\right)^n$ for $t< -\log(1-p) $
and the MGF for the Poisson distribution is $M_Y(t) = e^{\lambda(e^{t}-1)}$ 
Now since $\lim_{n \to 0} M_X(t) = \lim_{n \to 0} \left(\frac{p}{1-(1-p) e^t}\right)^n = 1$ is equal to $M_Y(t)|_{\lambda=0} = e^{0}=1$ 
would suggest that the Poisson distribution with $\lambda=0$ is the limiting distribution of the negative binomial distribution as $n$ approaches zero. The pmf for the Poisson distribution with $\lambda=0$ is $P(0)=1$ and $P(k)=0$ for $k>0$

**Comment by mdhaber** [comment]
@rkern do you have an opinion about whether $n=0$ should be considered to be in the domain of the negative binomial/Polya distribution?

**Comment by pbrod** [comment]
> If we think about the definition in terms of successes and failures, we can get into trouble. For instance, MathWorld's definition, which leads to the formula given in SciPy's documentation, is: ![image](https://github.com/scipy/scipy/assets/6570539/f330a02a-e436-4fda-9d77-948fe81734a2) You can see why it is problematic to talk about the probability of −1 successes in 0 trials and success on the 0th trial.

I don't see this as problematic. If you have zero ($x+r-1=0$) trials, you have zero ($x=0$) failures  and zero ($r-1=0$) successes , so logically the negative binomial distribution should give  $P(0)=1$ and $P(x)=0$ for $x>0$.

**Comment by josef-pkt** [comment]
does scipy already have other distributions that allow for  parameters that result in degenerate (single masspoint) distribution?
binom ?

For discrete distributions it might be ok because we already have masspoints.
However, var -> 0 and skew and kurtosis divide by zero.

same if p -> 1

**Comment by mdhaber** [comment]
@pbrod The situation of 0 success and 0 failures in 0 trials makes sense. That corresponds with $r=1$; the success is on the first trial. That's not what my comment was about. 

If you plug in $r=0$ and $x=0$, this corresponds with -1 successes in -1 trials, and success on the zeroth trial. How is that not problematic?

**Comment by josef-pkt** [comment]
nbinom is defined by using gamma function and not `choose`.
We have an extension to continuous parameters. 

n -> 0 works for n > 0, just plugging into current code (I'm currently still on scipy 1.7.3)

**Comment by mdhaber** [comment]
Yes, we're aware. The statement is about interpretation in terms of integer successes in failures.

**Comment by josef-pkt** [comment]
that interpretation also doesn't work if we have e.g. n=1.5 or n = 10.8750674865375

**Comment by mdhaber** [comment]
But those are defined in a reference. $r=0$ is not, or if it is, that hasn't been presented yet.

**Comment by josef-pkt** [comment]
I just did a google search and did not find anything for the degenerate case.

to the interpretation
wikipedia https://en.wikipedia.org/wiki/Negative_binomial_distribution#Alternative_formulations
has several interpretations

from nbinom docstring
"Negative binomial distribution describes a sequence of i.i.d. Bernoulli trials, repeated until a predefined, non-random number of successes occurs."

n is the required number of successes
if n=0, we stop after finding zero successes, i.e. we stop immediately, and then the number of failures k is also 0, i.e. single masspoint at zero.

the problem is that if both n->0 and p->0, then they go in opposite direction and it is not clear whether it collapses to the masspoint at zero, but p > 0 is required in code and docstring.

**Comment by mdhaber** [comment]
OK, I can see it making some sense according to that wording, although I would usually trust what I find on Mathworld more than SciPy documentation. Perhaps @ev-br has an opinion based on https://github.com/scipy/scipy/pull/6562? I don't mean to get in the way of others submitting/reviewing/merging PRs on this topic.

**Comment by ev-br** [comment]
I don't think I've anything useful to say here, sorry.

**Comment by pbrod** [comment]
Experimenting with the nbinom also indicates that the this is the correct limiting behavior of the negative binomial distribution as shown here:

    >>> import scipy.stats as ss
    >>> import numpy as np

    >>> ss.nbinom(n=0.1, p=0.5).pmf(np.arange(3))
    array([0.93303299, 0.04665165, 0.0128292 ])
    >>> ss.nbinom(n=1e-10, p=0.5).pmf(np.arange(3))
    array([1.00e+00, 5.00e-11, 1.25e-11])
    >>> ss.nbinom(n=1e-250, p=0.5).pmf(np.arange(3))
    array([1.00e+000, 5.00e-251, 1.25e-251])

    >>> ss.nbinom(n=0.1, p=0.01).pmf(np.arange(3))
    array([0.63095734, 0.06246478, 0.03401207])
    >>> ss.nbinom(n=1e-10, p=0.01).pmf(np.arange(3))
    array([1.0000e+00, 9.9000e-11, 4.9005e-11])
    >>> ss.nbinom(n=1e-250, p=0.01).pmf(np.arange(3))
    array([1.0000e+000, 9.9000e-251, 4.9005e-251])

    >>> ss.nbinom(n=1e-250, p=1e-250).pmf(np.arange(3))
    array([1.e+000, 1.e-250, 5.e-251])

    >>> ss.nbinom(n=0.1, p=1).pmf(np.arange(3))
    array([1., 0., 0.])
    >>> ss.nbinom(n=1e-250, p=1).pmf(np.arange(3))
    array([1., 0., 0.])

**Comment by pbrod** [comment]
The negative binomial distribution is a special case of the Katz distribution. Commonly also known as the Panjer or 
[(a,b,0) class of distributions](https://en.wikipedia.org/wiki/(a,b,0)_class_of_distributions).

Katz studied the system of discrete distributions that satisfy the relationship

$P_K(k+1) / P_K(k) = \frac{\alpha + \beta k}{k+1}  \quad \text{for} \quad k = 0, 1, 2, \ldots \qquad (1)$

where  $P_K(0)=(1-\beta)^\frac{\alpha}{\beta}$, $\alpha \ge 0$ and $\beta<1$ and showed that there are only three solutions to Eq.1:

$\text{NegativeBinomial}(n=\alpha/\beta, p = 1-\beta) \quad \text{for} \quad 0< \beta < 1$,
$\text{Poisson}(\lambda=\alpha) \quad \text{for} \quad \beta = 0 $
or
$\text{Binomial}(n=-\alpha/\beta, p = \beta/(\beta-1)) \quad \text{for} \quad \beta < 0 $.

Katz originally truncated $P_K(k+j)$ to zero whenever  $\alpha + \beta k<0$. However Washburn [(1996,](https://www.researchgate.net/publication/235203117_Katz_Distributions_with_Applications_to_Minefield_Clearance) [2006)](https://www.jstor.org/stable/43941048)) recognized that $-\alpha/\beta$ 
actually must be an integer greater or equal to zero for the Binomial case.

The [moment generating function ](https://en.wikipedia.org/wiki/Moment-generating_function)(MGF) for the Katz distribution is $M_X(t) = \left(\frac{1-\beta}{1-\beta e^t}\right)^{\alpha/\beta}$ for $t< -\log(\max(\beta, 0))$.
For $\beta \to 0$ it is interpreted as $\lim_{\beta \to 0}M_X(t)  = e^{\alpha(e^t-1)}$ 

Now for $\beta \ne 0$ the limit of the MGF as $\alpha$ goes to zero is $\lim_{\alpha \to 0} M_X(t) = \lim_{\alpha \to 0} \left(\frac{1-\beta}{1-\beta e^t}\right)^{\alpha/\beta} = 1$
For $\beta = 0$ the limit of the MGF as $\alpha$ goes to zero is $\lim_{\alpha \to 0, \beta=0} M_X(t) = \lim_{\alpha \to 0}  e^{\alpha(e^t-1)}=e^0= 1$
This result proves that the Katz distribution as $\alpha \xrightarrow[]{} 0$ is the same regardless of what the value of $\beta$ is.

Thus the pmf of the Katz distribution with $\alpha=0$  is $P_K(0)=1$ and $P_K(k)=0$ for $k>0$.
Consequently the negative binomial distribution for $n=0$ should have the same pmf, which supports the conclusion in
https://github.com/scipy/scipy/issues/18974#issuecomment-1655357122.

**Comment by steppi** [comment]
> @steppi This issue proposes to define the PMF of the negative binomial distribtion with n = 0 successes as a point mass with probability 1 at k = 0 failures. I think this holds even for probability of success p = 0 . Do you think this is something we would want to adjust in the wrappers, would you ask Boost for the change, or neither?

If this is worth doing, my vote is for adjusting it in the wrappers. There's nothing wrong with Boost's convention, and it matches their [documentation](https://beta.boost.org/doc/libs/1_35_0/libs/math/doc/sf_and_dist/html/math_toolkit/dist/dist_ref/dists/negative_binomial_dist.html).

*"For k + r Bernoulli trials each with success fraction p, the negative_binomial distribution gives the probability of observing k failures and r successes with success on the last trial. The negative_binomial distribution assumes that success_fraction p is fixed for all (k + r) trials."*

It's also a lot easier in general for Boost users to write their own wrappers if they care about the `r = 0` case, compared to adjusting a `scipy` distribution.

**Comment by mdhaber** [comment]
Right, I know that their documentation currently mentions that the trial ends with a success, and so the change would not be consistent with it as written. Of course, the existing documentation does not support non-integer `n`, either. 

Ok, so we should not trouble Boost with this. I can say that I do not think it's worth changing in `stats` because it would add overhead in all cases to support this edge case. Your call @steppi whether it should be done in `special`.
